### ISSUE TITLE
Async Request Feature Request

### ISSUE THREAD

**Issue by noboomu** [jooby-project/jooby#173]
No sure if I'm simply not finding it in the docs, but async request processing would be great. If Jooby had support for async and modules for ebean and akka I know a lot of folks that would be happy dropping the playframework.

**Comment by jknack** [comment]
can you help me to define it? in terms of API.

Today, for simplicity and easy to use, every single request across servers (netty, jetty, undertow) can block.

**Comment by noboomu** [comment]
Let me take a deeper look :)

**Comment by jknack** [comment]
think https://github.com/jdeferred/jdeferred is a good option

```java
{
  get("/async", req -> {
   return new DeferredObject().promise().then(...).done(() -> {
     return ...;
   });
  });
}
```

**Comment by noboomu** [comment]
I've never heard of jdeferred before, looks great! I had been thinking of https://github.com/advantageous/qbit, https://github.com/puniverse/quasar, or https://github.com/ReactiveX/RxJava but jdeferred looks good!

**Comment by jknack** [comment]
still want to keep Jooby simplicity, so I'm thinking on introduce a "server mode" feature:

default mode:

```java
{
  get("/", () -> "...");

  get("/", () -> Deferred.promise(() -> "..."))
}
```

The previous two route are exactly the same, there is no difference between them in **default mode**.

nio mode:

```java
{
  mode(NIO);

  get("/", () -> "...");

  get("/", () -> Deferred.promise(() -> "..."));
}
```

In NIO mode, the first route will run in the NIO thread, but the second route  don't. If route need to block... a promise/deferred object is required.

For Promise? I think it would be nice to have a Deferred object (like jquery or jdeferred) and use ```CompletableFuture``` from Java 8. This way, it provides a clean and rich API without any external dependency (RxJava, JDeferred, etc..)

Thoughts?

**Comment by noboomu** [comment]
I like it. Especially the limit on external libs :)

**Comment by paulovictorv** [comment]
Me again.

I have a promise background coming from my days of AngularJs deving, and I find the CompletableFuture API very hard to understand. Maybe it's just me, but I haven't being able to use it at all on my projects. The most powerful Promise feature is the ability to chain promises together, something that is not very well documented on the CompletableFuture API. I'm completely open to be told otherwise, though.

The guys that implemented Bluebird (for Javascript) did an awesome job in implementing the Promise spec, and JDeferred is the closest representative on the Java side.

I think that if Jooby is going to support async requests, JDeferred is the way to go.

**Comment by jknack** [comment]
agreed that CompletableFuture API is hard to follow and use... did something which basically wrap a completable future and offers a short and concise API like deferred from jquery.

JDeferred looks good, but for example it let you specify (or actually you must specify) the type of the rejected value. For me, the rejected value should always be an Exception...

another thing to keep in mind is how jooby handle filters, so can't just said: "let's return a jdeferred/callable/completablefuture"...


more and more I think, feel jooby should provides a more clean, simple and easy to use API for async processing (like servlet 3.1 and/or jersey)


```java
{
  get("/", req -> {
     req.async(ctx -> {
        // cpu intensive or blocking calls
        ctx.end();
     });
  });
}
``` 

or via your favorite executor:

```java
{
  get("/", req -> {
     req.async(ctx -> {
        // cpu intensive or blocking calls
        ctx.end();
     }, executor);
  });
}
```

This also allow us to pick/use our favorite async library:

```java
{
  get("/", req -> {
     req.async(ctx -> {
        // jdeferred, akka, whatever...
        ctx.end();
     }, executor);
  });
}
```

Thoughts?

**Comment by paulovictorv** [comment]
This looks great actually. It doesn't tie you down to a specific async lib
"vendor", and stays clear from the application code async handling.

The opposite of this is exactly why I hate Play Framework: it ties you down
too much, specially around Async stuff.

Em segunda-feira, 24 de agosto de 2015, Edgar Espina <
notifications@github.com> escreveu:

> agreed that CompletableFuture API is hard to follow and use... did
> something which basically wrap a completable future and offers a short and
> concise API like deferred from jquery.
>
> JDeferred looks good, but for example it let you specify (or actually you
> must specify) the type of the rejected value. For me, the rejected value
> should always be an Exception...
>
> another thing to keep in mind is how jooby handle filters, so can't just
> said: "let's return a jdeferred/callable/completablefuture"...
>
> more and more I think, feel jooby should provides a more clean, simple and
> easy to use API for async processing (like servlet 3.1 and/or jersey)
>
> {
>   get("/", req -> {
>      req.async(ctx -> {
>         // cpu intensive or blocking calls
>         ctx.end();
>      });
>   });
> }
>
> or via your favorite executor:
>
> {
>   get("/", req -> {
>      req.async(ctx -> {
>         // cpu intensive or blocking calls
>         ctx.end();
>      }, executor);
>   });
> }
>
> This also allow us to pick/use our favorite async library:
>
> {
>   get("/", req -> {
>      req.async(ctx -> {
>         // jdeferred, akka, whatever...
>         ctx.end();
>      }, executor);
>   });
> }
>
> Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jooby-project/jooby/issues/173#issuecomment-134405149>
> .
>


-- 
Paulo Victor Vieira de Melo
Graduando em Ciência da Computação
Lattes : http://lattes.cnpq.br/7418993216312941
Currículo StackOverflow Carrers:
http://careers.stackoverflow.com/paulovictorv
UFAL - IC

**Comment by jknack** [comment]
hey, me again,what do you all think about these two APIs example:

```java
get("/async1", async(req -> {
     return ...;
}));
```

vs

```java
get("/async1", req -> {
     return ...;
}).async();
```

want to avoid callback hell by going async... Jooby API is clean and easy to use, want to keep that when going async

first example enable async via a custom/specific handler:

```java
async(req, -> ...)
```

2nd example will do it via route definition

```java
get("async", req -> {
}).async();
```

2nd option also work on group of routes:

```java
use("/pets")
   .get(...)
   .post(...)
   ...
   .async();
```

Also, not sure what name is best for describing the async processing: ```async``` or ```dispatch```

What do you ALL think?

**Comment by paulovictorv** [comment]
I have mixed feelings. I like them both.

Is it possible to support both syntaxes?

**Comment by jknack** [comment]
Hi,

 I'm going to drop the: ```mode(NIO)``` mode, when I visualize this thought it will be nice to allow application code to run in the IO thread and let developer to go async explicitly... this sounds great at first and seems to be a super powered feature... but makes everything complex and error prone.

 I'm going to drop such feature and leave it as it is today:

1. **Async**
2. Application code runs in a **worker thread**
3. A **worker thread** can block

What then?

Next release of Jooby will come with a ```Deferred``` implementation. Jooby deferred will be very very simple, it will have just one method: ```set````

The use of deferred will detach the worker thread and put the request in async mode.

Advanced/Async users will be able to use the async library of their choice (RxJava, Akka, JDeferred, etc...) and build 100% async apps.

Here are some examples:

* Using the good and old executor:

```java
{
    ExecutorService executor = ...;

    get("/", deferred(result -> {
      executor.execute(() -> {
        result.set("async");
      });
    }));
}
```

* RxJava:

```java
{
   get("/rx", deferred((req, result) -> {
      Observable.<List<String>> create(s -> {
        s.onNext("async");
        s.onCompleted();
      }).subscribeOn(Schedulers.computation())
        .subscribe(result::set, result::set);
  }));
}
```

From a MVC controller:

```java
@Path("/")
public class Resource {

   public Deferred async() {
     return new Deferred(result -> {
        executor.execute(() -> {
          result.set("async");
        });
     });
   }
}
```

Cool?

**Comment by paulovictorv** [comment]
That looks cool to me, it avoids vendor lock in :)

**Comment by jknack** [comment]
0.10.0 is out: http://localhost:4000/doc/#async-routes

**Comment by noboomu** [comment]
Fantastic. Is there a mechanism to call async from MVC routes? Hoping to get swagger support for the async endpoints.

**Comment by jknack** [comment]
yea:

```
public class Mvc {

  @Pat("/") @GET
   public Deferred async() {
      return new Deferred(() -> {
        ...
      });
  }
}
```

and Swagger will just works, nothing has been changed there :)

**Comment by noboomu** [comment]
Perfect, migrating now :)

••••••••••••••••••
